### PR TITLE
Continue bootstrapmethods parsing after unsupported entry

### PR DESCRIPTION
**Pull request by mgudemann**
Before it was possible to keep some `BootstrapMethods` entries unparsed, after a
currently unsupported was found. This lead to the situation that JBMC ended up
with a wrong file offset which caused problems in the case of class file
attributes being present after the `BootstrapMethods` attribute.

This was not observed using the Oracle or OpenJDK compilers, but currently only
on the Eclipse compiler for Java.

### CODE REVIEW

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 1]
```diff
@@ -1803,7 +1803,7 @@ java_bytecode_parsert::parse_method_handle(const method_handle_infot &entry)
 void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
 {
   u2 num_bootstrap_methods = read_u2();
-  for(size_t i = 0; i < num_bootstrap_methods; i++)
+  for(size_t j = 0; j < num_bootstrap_methods; j++)
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_parser.cpp]: Why has this changed from `i` to `j`

Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp]: Because it was colliding with another loop, perhaps a good argument for a better name like `boostrap_method_index`?

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 2]
```diff
@@ -1868,72 +1868,81 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
...
-                << "\n  interface type is "
-                << id2string(pool_entry(interface_type_argument.ref1).s)
-                << "\n  method type is "
-                << id2string(pool_entry(method_type_argument.ref1).s) << eom;
+        const pool_entryt &interface_type_argument =
+          pool_entry(argument_index1);
+        const pool_entryt &method_handle_argument = pool_entry(argument_index2);
+        const pool_entryt &method_type_argument = pool_entry(argument_index3);
+
+        if(
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_parser.cpp]: It is a bit hard to follow these nested if/else statements, consider changing to:
```c++
if(! /* validity check 1 */) 
{
  report_error1();
  continue;
}
if(! /* validity check 2 */) 
{
  report_error2();
  continue;
}
// do the real work!
```

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+CORE
+StaticMethodRef.class
+--function StaticMethodRef.test2
+^EXIT=0
+^SIGNAL=0
+--
+--
+Tests that it doesn't crash when failing to read a `BootstrapMethods` entry and
```
Comment (thk123) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: Is this compiled with eclipse compiler? If so probably worth making a note of that here to stop people inadvertently updating it.

File: src/java_bytecode/java_bytecode_parser.cpp (hunk 2 above)
Comment (majakusber) [reply src/java_bytecode/java_bytecode_parser.cpp]: +1 Some commentary on the individual cases would make it easier to follow too.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+CORE
+StaticMethodRef.class
+--function StaticMethodRef.test2
```
Comment (majakusber) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: Why this method is tested instead of `Smr`? Wouldn't `Smr` do the same job, with lambdas being ignored but parsed?

**Review by majakusber — CHANGES_REQUESTED** [review]
The if-else statements are nested now which makes them hard to read. It would be useful for future work on lambdas to make it more readable.

**Comment by thk123** [comment]
I've done as suggested and replaced the nested ifs with `if(!valid) continue;` style code as well as pulling out the error producing code into a separate function to reduce duplication. I've also tried to document each of the errors, so @svorenova could you see if they make sense? Will consider the tests next.

**Comment by majakusber** [comment]
@thk123 Yes, looks very good, thank you! Will wait for the test to approve.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc (hunk 3 above)
Comment (thk123) [reply regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: It appears to be - will update docs

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc (hunk 4 above)
Comment (thk123) [reply regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: I agree and the test passes setting to the the `Smr` so will change to that but @mgudemann when you're back if you could comment on whether this was the right thing to do.

**Comment by thk123** [comment]
@svorenova Adjusted the tests

File: regression/cbmc-java/lambda2/StaticMethodRef.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import java.util.function.Function;
+import java.util.function.BiFunction;
+
+public class StaticMethodRef{
+ public Integer test2(){return 2;}
```
Comment (majakusber) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: This can be removed now it seems.

**Review by majakusber — APPROVED** [review]
Looks good, just one line in the test to remove but non-blocking.

**Review by thk123 — APPROVED** [review]
I've applied my comments, so I'm happy, probably good to get a second review from a different "senior" e.g. @allredj ?

**Comment by allredj** [comment]
@thk123 OK I'll take a look.

File: regression/cbmc-java/lambda2/StaticMethodRef.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+   BiFunction<String, Integer,  Integer> func3  = Integer::valueOf;
+
+   if (ctr.equals(func1.apply(9)))
+       return func1.apply(9);
+
+   if(ctr.equals(func2.apply("8")))
+       return func2.apply("8");
+
+   if(ctr.equals(func3.apply("0111", 2)))
```
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Nit: Java format is not uniform across the file.

File: regression/cbmc-java/lambda2/StaticMethodRef.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.function.BiFunction;
+
+public class StaticMethodRef{
+ public Integer Smr(Integer ctr) {
+   Function<Integer, Integer> func1  = Integer::valueOf;
+
+   // Uses  Integer.valueOf(String)
+   Function<String, Integer> func2  = Integer::valueOf;
+
+   BiFunction<String, Integer,  Integer> func3  = Integer::valueOf;
```
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Nit: extra spaces in the generic signature.

File: regression/cbmc-java/lambda2/StaticMethodRef.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+       return func1.apply(9);
+
+   if(ctr.equals(func2.apply("8")))
+       return func2.apply("8");
+
+   if(ctr.equals(func3.apply("0111", 2)))
+       return func3.apply("0111", 2);
+
+   return ctr;
+ }
```
Comment (allredj) [inline regression/cbmc-java/lambda2/StaticMethodRef.java]: Bad indent

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 9]
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
...
+    if(
+      !(interface_type_argument.tag == CONSTANT_MethodType &&
+        method_handle_argument.tag == CONSTANT_MethodHandle &&
+        method_type_argument.tag == CONSTANT_MethodType))
+    {
+      debug() << "format of BootstrapMethods entry not recognized: arguments "
+                 "wrong type"
+              << eom;
+      store_unknown_method_handle(
+        parsed_class, bootstrap_method_index, std::move(u2_values));
```
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp:1906]: Question for my understanding: Why do we want to keep these unknown handles instead of just skipping the?

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 10]
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
...
+            << id2string(pool_entry(method_type_argument.ref1).s)
+            << eom;
+    parsed_class.lambda_method_handle_map[{parsed_class.name, bootstrap_method_index}] =
+      *lambda_method_handle;
   }
 }
+
+/// Creates an unknown method handle and puts it into the parsed_class
+/// \param parsed_class: The class whose bootstrap method handles we are using
+/// \param bootstrap_method_index: The current index in the boostrap entry table
```
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: `boostrap` -> `bootstrap`

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 11]
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
...
+    lambda_method_handle->u2_values = std::move(u2_values);
+    debug() << "lambda function reference "
+            << id2string(lambda_method_handle->lambda_method_name)
+            << " in class \"" << parsed_class.name << "\""
+            << "\n  interface type is "
+            << id2string(pool_entry(interface_type_argument.ref1).s)
+            << "\n  method type is "
+            << id2string(pool_entry(method_type_argument.ref1).s)
+            << eom;
+    parsed_class.lambda_method_handle_map[{parsed_class.name, bootstrap_method_index}] =
```
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: Nit: Don't we have an 80-char limit?

Comment (allredj) [reply src/java_bytecode/java_bytecode_parser.cpp]: The way we access this map looks a bit unnatural. Why does name of `parsed_class` have to be an element of the map entry, that itself belongs to the same class. Obviously the name will always be identical for a given `parsed_class` or am I missing something? Could we have a wrapper method in `java_bytecode_parse_treet::classt` that cares about adding an element to the map instead of accessing the map directly?

File: src/java_bytecode/java_bytecode_parser.cpp [hunk 12]
```diff
@@ -1820,122 +1827,137 @@ void java_bytecode_parsert::read_bootstrapmethods_entry(classt &parsed_class)
...
+
+    const pool_entryt &interface_type_argument =
+      pool_entry(interface_type_index);
+    const pool_entryt &method_handle_argument = pool_entry(method_handle_index);
+    const pool_entryt &method_type_argument = pool_entry(method_type_index);
+
+    if(
+      !(interface_type_argument.tag == CONSTANT_MethodType &&
+        method_handle_argument.tag == CONSTANT_MethodHandle &&
+        method_type_argument.tag == CONSTANT_MethodType))
```
Comment (allredj) [inline src/java_bytecode/java_bytecode_parser.cpp]: Nit: I find conditions of the form:
```
interface_type_argument.tag != CONSTANT_MethodType ||
method_handle_argument.tag != CONSTANT_MethodHandle ||
method_type_argument.tag != CONSTANT_MethodType
```
easier to read

**Review by allredj — APPROVED** [review]
Looks fine as far as I can tell. I have a couple of questions.

File: regression/cbmc-java/lambda2/test_no_crash_ecj.desc [hunk 13]
```diff
@@ -1,7 +1,9 @@
 CORE
 StaticMethodRef.class
 --function StaticMethodRef.Smr
-^EXIT=0
+VERIFICATION FAILED
+EXIT=10
+SIGNAL=0
```
Comment (allredj) [inline regression/cbmc-java/lambda2/test_no_crash_ecj.desc]: That's a bit confusing for whoever will see this test fails once lambdas start working. Could we add a comment saying what specifically we are testing here?

File: src/java_bytecode/java_bytecode_parser.cpp (hunk 9 above)
Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp:1906]: I think because the `invokedynamic` refers to an index in the bootstrap table which would be invalid if we didn't populate all the entries .

File: src/java_bytecode/java_bytecode_parser.cpp (hunk 11 above)
Comment (thk123) [reply src/java_bytecode/java_bytecode_parser.cpp]: Yes I agree  tagging @mgudemann to specifically review what I've done here but I agree this info is redundant and a method will make updating this in the future simpler - will do.

**Comment by thk123** [comment]
@allredj Addressed all comments, though have spun out as a separate ticket to make the bootstrap methods map itself private as a bit more involved

**Comment by allredj** [comment]
@thk123 thanks!
